Check if file exists before sorting them into a list

kush-gupt opened the pull request:
A dangling symlink can cause `ramalama ls` to fail with:
```
$ ramalama ls            
Error: [Errno 2] No such file or directory: 'ollama/qwen:1.8b
```

So checking to see if the file exists before appending it into a list to be sorted fixes this:
```
$ ramalama ls
NAME                       MODIFIED     SIZE    
ollama://starcoder2:3b     1 hour ago   1.59 GB 
ollama://deepseek-r1:32b   12 hours ago 18.49 GB
ollama://qwen2.5-coder:32b 12 hours ago 18.49 GB
```

Root cause could be solved by creating a hard link instead of symbolic from to ollama model blob to the ramalama blob store

## Summary by Sourcery

Bug Fixes:
- Fix an error where a dangling symlink could cause `ramalama ls` to fail.

kush-gupt commented:
Fixes the issue reproduced here: https://github.com/containers/ramalama/pull/782#issuecomment-2650930131

kush-gupt commented:
In the context of where the other inodes are located, it _should_ be just the ollama and hf caches right? So the user would have to ollama/hf rm them separately to completely wipe the model from their filesystem but that's also what we'd expect right? 

Scenario I see is: User pulls some model with ramalama, it's nice to load from cache but it's still on the source cache/tooling to manage the original model?

rhatdan commented:
I was thinking of creating a hard link to a file in the same destination to where ramalama pull ollaman:foobar puts it now.

So you would still have a symlink, but instead of a symlink to the file under OLLAMA it would be a symlink to a hardlink file under $HOME/.local/share/ramalama

rhatdan commented:
With this change does `ramalama rm MODEL` with broken symlink report MODEL does not exists?  I want ramalama rm MODEL to remove broken symlink.

kush-gupt commented:
ramalama rm works to remove the broken symlink:
```
$ ramalama rm qwen:1.8b
Untagged: qwen:1.8b
Deleted: sha256:1296b084ed6bc4c6eaee99255d73e9c715d38e0087b6467fd1c498b908180614
Deleted: sha256:9ece4a97bfb61bdb539531db5584fa119ad55684281d8a2d864339ae3fdd6c15

$ ls $HOME/.local/share/ramalama/models/ollama/
deepseek-r1:32b		qwen2.5-coder:32b	starcoder2:3b
```

ericcurtin commented:
> With this change does `ramalama rm MODEL` with broken symlink report MODEL does not exists? I want ramalama rm MODEL to remove broken symlink.

I'm not sure there is a case where broken symlink happens. I suspect @kush-gupt just created a dangling symlink already. The problem with hardlinks is the Ollama structure:

```
                  ----> modela:latest
sha256:somechecksum 
                  ----> modela:8b
```

8b and latest are the same file, both just symlinks for de-duplication purposes.

If we make things hardlinks, we have to do a lot of searching for the other linked hardnames, it will be expensive. We don't really gain the automatic reference counting cleanup capabilities of hardlinks because we don't refer to models by their checksum (and we have to find that checksum somehow, with symlink it's easy because you are pointed to that SHA, when it's a hardlink we aren't pointed to any sha)

ericcurtin commented:
IIRC the "ramalama rm" does a decent job of trying to delete all orphaned things, which is why here we see two sha256's deleted in @kush-gupt 's output

kush-gupt commented:
Indeed, I had that broken symlink created after I did `ollama rm qwen:1.8b` after I had already done a `ramalama pull qwen:1.8b`

rhatdan commented:
Ok I think I am ok with just symlink as long as RamaLama handles it.  I think we also create a symlink to a model stored as a file.

file:///tmp/model.gguf for example.

rhatdan commented:
I would like to see a test for this though.

Create a model with file://$TEMPDIR/model.gguf  Show it in list.  Remove $TEMPDIR/model.gguf
Now see if the list sees it.  Then attempt to remove it and make sure the rm succeeds.

kush-gupt commented:
Added a call to remove if a broken symlink is detected when listing models:
```
$ ollama ls
NAME                                   ID              SIZE      MODIFIED     
nomic-embed-text:latest                0a109f422b47    274 MB    8 months ago  
  
$ ramalama ls 
NAME                       MODIFIED     SIZE    
ollama://starcoder2:3b     7 hours ago  1.59 GB 
ollama://deepseek-r1:32b   17 hours ago 18.49 GB
ollama://qwen2.5-coder:32b 17 hours ago 18.49 GB

$ ramalama pull nomic-embed-text:latest 

$ ollama rm nomic-embed-text:latest 
deleted 'nomic-embed-text:latest'

kugupta@kugupta-mac ramalama % ramalama ls                          
Broken symlink found in: /Users/kugupta/.local/share/ramalama/models/ollama/nomic-embed-text:latest 
Attempting removal
Untagged: nomic-embed-text:latest
Deleted: sha256:970aa74c0a90ef7482477cf803618e776e173c007bf957f635f1015bfcfef0e6
NAME                       MODIFIED     SIZE    
ollama://starcoder2:3b     7 hours ago  1.59 GB 
ollama://deepseek-r1:32b   17 hours ago 18.49 GB
ollama://qwen2.5-coder:32b 18 hours ago 18.49 GB
```

rhatdan commented:
LGTM

maxamillion reviewed (approved):
LGTM
